docs: Blog post for GA of container generator (#577)
slsa-framework/slsa: docs/_posts/2023-02-01-slsa-github-workflows-container-ga.md

---
title: "General availability of SLSA 3 Container Generator for GitHub Actions"
author: "Asra Ali, Ian Lewis, Laurent Simon"
is_guest_post: false
---

> Today, we are announcing the general availability of the [SLSA 3 Container Generator for GitHub Actions](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container) starting with v1.4.0. This free tool allows any GitHub project to produce SLSA level 3 compliant provenance statements so users can verify the origin of container images they use. While previous tools allowed users to generate provenance for file artifacts, the Container Generator is able to support container ecosystems. It does this by allowing provenance statements to be distributed alongside your images in a container registry and integrate directly with [Sigstore](https://www.sigstore.dev/)-compatible tooling for inspection and verification.

Reviewer: ```suggestion
Today, we are announcing the general availability of the [SLSA 3 Container Generator for GitHub Actions](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container) starting with v1.4.0. This free tool allows any GitHub project to produce SLSA level 3 compliant provenance statements so users can verify the origin of container images they use. While previous tools allowed users to generate provenance for file artifacts, the Container Generator is able to support container ecosystems. It does this by allowing provenance statements to be distributed alongside your images in a container registry and integrating directly with [Sigstore](https://www.sigstore.dev/)-compatible tooling for inspection and verification.
```

In only a few short years since container management platforms like Docker and Kubernetes were released, container images have become one of the most popular ways to deploy software. Containers allow their users to deploy software in a more flexible way without depending on language specific packaging or deployment mechanisms. Many public repositories of container images, like [Docker Hub](https://hub.docker.com/), allow you to upload pre-packaged images that can be used out of the box or customized by users as base images.

> However, it can be a huge challenge for users to verify the integrity of your images. Various reports from [Aqua Security's Team Nautilus](https://blog.aquasec.com/supply-chain-threats-using-container-images) and the [Sysdig Threat Research Team](https://sysdig.com/blog/analysis-of-supply-chain-attacks-through-public-docker-images/) show that container images containing malicious payloads are common. How can users of your image be certain of how the images you use were built? How can they be sure they haven't been modified to include malicious payloads?

Reviewer 2: This paragraph is a little awkward to read. I've made a couple of suggestion.

```suggestion
However, it can be a huge challenge for users to verify the provenance of your images. Various reports from [Aqua Security's Team Nautilus](https://blog.aquasec.com/supply-chain-threats-using-container-images) and the [Sysdig Threat Research Team](https://sysdig.com/blog/analysis-of-supply-chain-attacks-through-public-docker-images/) show that container images which include malicious payloads are common. How can users of your image be confident of how the images you use were built? How can they be sure they haven't been modified to include malicious payloads?
```

One solution is including [SLSA provenance](/provenance/) with your container image. SLSA provenance is metadata about your build that includes information that strongly links your artifacts to the build system and source code that was used. SLSA provenance can build trust in your images by helping your users verify that the images come from you and your source repositories.

The Container Generator's reusable workflow can help you to achieve [SLSA level 3](/spec/v0.1/levels) for your project by automatically generating provenance from your existing GitHub Actions workflows to generate images. This can be done by simply adding a new job step to call the generator's reusable workflow, which generates and signs the provenance in an isolated job separate from your image's build step for added security. Popular projects like [Kyverno](https://kyverno.io/docs/) are already using this workflow to generate provenance for their container image releases.

## Generating Provenance

> Under the hood, the Container Generator's reusable workflow uses [Sigstore](https://sigstore.dev/)'s [cosign](https://github.com/sigstore/cosign) to sign the provenance. This provenance is combined with the cryptographic signature into something called an attestation. This attestation is then uploaded to the container registry. Using cosign to sign the provenance allows the generated attestation to integrate well with sigstore's ecosystem of tools for discovery and verification.

Reviewer 2: Nit: sigstore->Sigstore
```suggestion
Under the hood, the Container Generator's reusable workflow uses [Sigstore](https://sigstore.dev/)'s [cosign](https://github.com/sigstore/cosign) to sign the provenance. This provenance is combined with the cryptographic signature into something called an attestation. This attestation is then uploaded to the container registry. Using cosign to sign the provenance allows the generated attestation to integrate well with Sigstore's ecosystem of tools for discovery and verification.
```

> However, unlike simply signing your container directly, using the Container Generator as a [GitHub Actions reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) generates the signed provenance in an isolated build job. This protects the provenance from being modified by malicious code in the build itself and enables you to meet the [Non-Falsifiable requirements](https://slsa.dev/spec/v0.1/requirements#non-falsifiable) for [SLSA level 3](https://slsa.dev/spec/v0.1/levels). You can find out more about the specifics in our [Technical Design](https://github.com/slsa-framework/slsa-github-generator/blob/main/SPECIFICATIONS.md).

Reviewer 3: What about clarifying "simply signing your attestation on the container directly"?

Not sure though - I like the clear cut of this wording, but it also ignores the difference that simple signing doesn't explicitly give you the like to the source / metadata

Author: Right, maybe it was a mistake to bring it up but in the container ecosystem a lot of folks already know about cosign and use it to sign without having full provenance.

Also, source <-> artifact linking is possible via the cert metadata so astute folks would probably notice that.

> However, unlike simply signing your container directly, using the Container Generator as a [GitHub Actions reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) generates the signed provenance in an isolated build job. This protects the provenance from being modified by malicious code in the build itself and enables you to meet the [Non-Falsifiable requirements](/spec/v0.1/requirements#non-falsifiable) for [SLSA level 3](/spec/v0.1/levels). You can find out more about the specifics in our [Technical Design](https://github.com/slsa-framework/slsa-github-generator/blob/main/SPECIFICATIONS.md).

Reviewer 2: ```suggestion
However, unlike simply signing your container directly with cosign, using the Container Generator as a [GitHub Actions reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) generates the signed provenance in an isolated build job. This protects the provenance from being modified by malicious code in the build itself and enables you to meet the [Non-Falsifiable requirements](/spec/v0.1/requirements#non-falsifiable) for [SLSA level 3](/spec/v0.1/levels). You can find out more about the specifics in our [Technical Design](https://github.com/slsa-framework/slsa-github-generator/blob/main/SPECIFICATIONS.md).
```

```yaml
build:
  # ...
  steps:
    # ...
    - name: Build and push Docker image
      uses: docker/build-push-action@e551b19e49efd4e98792db7592c17c09b89db8d8 # v3.0.0
      id: build
      with:
        push: true
        tags: {% raw %}${{ steps.meta.outputs.tags }}{% endraw %}
        labels: {% raw %}${{ steps.meta.outputs.labels }{% endraw %}}
    # ...
```

### Step 2: Call the Container Generator to generate and upload SLSA provenance

With the outputs from image generation you can call the Container Generator reusable workflow. This step calls the container workflow to generate provenance, sign it, and push the resulting attestation to the container registry. An example looks like this:

```yaml
provenance:
  needs: [build]
  permissions:
    actions: read # for detecting the Github Actions environment.
    id-token: write # for creating OIDC tokens for signing.
    packages: write # for uploading attestations.

>   uses: slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@v1.4.0

Reviewer 3: should you mention which versions it's supported in somewhere above? e.g. `reusable workflow supported in versions 1.4+`

---
title: "General availability of SLSA3 Container Generator for GitHub Actions"
author: "Asra Ali, Ian Lewis, Laurent Simon"
is_guest_post: false
---

> Today, we are announcing the general availability of the [SLSA3 Container Generator for GitHub Actions](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container). This free tool allows any GitHub project to produce SLSA compliant provenance statements so users can verify the origin of container images they use.

Reviewer: Can you explain why this is needed and how it compares to the generic one we announced previously? What benefits does it have, and why would someone choose this over the other one?

Author: I kind of tried to change things up a bit and emphasize this workflow is for containers, without necessarily mentioning the others (almost all of our previous posts started with something like "previously we announced xxx, now we're releasing YY, because ZZ") but maybe that kind of intro is unavoidable.

Does the new text explain it adequately? do we need more explanation regarding support for storing the attestation in container registries or is it apparent from the rest of the article?

Reviewer: It doesn't address it for me. I find it very confusing to know what the difference is between a "generic generator" and "container generator". This whole intro talks about why containers and SLSA are important but not why the existing SLSA tooling doesn't apply. I would think that "generator" would cover everything, including but not limited to containers.

More talk about what makes this thing different than what came before is warranted, in my opinion. (I also think the old "Today we're announcing" is fine, no need to start with "Previously...")

Author: Right, so the benefits are 
1. Sigstore's attestation bundle format/specification
2. Support for uploading files to the Container registry

Both allow you to plug into the existing Sigstore-aware tooling like `cosign`, Kyverno, `policy-controller` etc. The generic tooling isn't really usable for containers since that ecosystem expects attestations to be distributed alongside the containers in the registry as opposed to on a GitHub release.

I can try to address this a bit but I'm not sure that mentioning the bundle format etc. is the right level of detail for the blog post. I'll try to better emphasize the integration with ecosystem tooling.

Author: I added some more context to the opening:
> With the Container Generator these provenance statements can be distributed alongside your images in a container registry and integrate directly with [Sigstore](https://www.sigstore.dev/)-compatible tooling for inspection and verification

I think this also ties in better with the later sections on inspection and verification as well. Does this better explain the need for the container generator?

Reviewer: That is an improvement, but it's still too weak. That statement is true of the Generic Generator too, since it "can be" used in that fashion as well, just with more work. Suggestion:

> This new Container Generator automatically distributes provenance alongside the container image in a registry in a format that integrates directly with [Sigstore](https://www.sigstore.dev/)-compatible tooling.

Author: I suppose it's possible that the generic generator could take the subject, generate the right sigstore bundle, and upload it to the container registry for the user. This is probably possible using sigstore libraries.

There are a couple downsides I see with regard to this.
1. Because cosign generally expects to generate the intoto statement (including subject) itself from a given predicate, in the case of the generic generator, users would need to get the exact subject format right including the full image names (index.docker.io/foo/bar) in order for tools like `cosign verify-attestation`, Kyverno, sigstore policy-controller to work. We could try to improve the UX by auto-detecting if the subject is containers to make this easier but that's more container-specific support in the generic generator.
2. We would need to add container registry support to the generic generator. This is not necessarily a bad thing per-se, because there are maybe use cases where users would want to upload non-container image attestations to a container registry, but it complicates the UX.
3. As an alternative to container registry support we could output a sigstore bundle that users could upload to the container registry themselves. This forces users to know how to upload to the container registry in exactly the right way. AFAIK cosign doesn't provide an out of the box way to do that that works with `cosign verify-attestation` etc.

In general, I don't see being able to provide a good UX to users without a lot of container specific code paths in the generic generator, and this is the reason we went with a separate workflow. What we really wanted to provide to users was a way to provide the image name and digest and have it "just work".

Reviewer: Sorry, I'm not questioning the value of the container generator or suggesting changes to it. Rather, I'm saying that the wording as written in this blog post does not do a great job at getting your point across. IIUC, your intention is to convey that this new generator makes some tasks easier that are much harder and error prone with the generic generator. So say that explicitly.

In particular:

- "These provenance statements" - I suspect you mean "Sigstore-compatible bundle files" which are specific to this new generator, not the generic in-toto statement files. But that's never explained anywhere, so most readers would assume you mean in-toto. Which is what the generic generator produces.
- "can be" - Most tools "can be" used to do things. As I said, the old one "can be" used as well. What you want to say is that it does it for you automatically.
- The info in your last response is probably not understood by most readers, so they won't appreciate how difficult it was to use a container with the generic generator. It would help to tell them.

Reviewer 4: @bobcallaway @olivekl - thoughts on how to better rephrase this ? Also, see Mark's comment below as well.

Reviewer 5: My 2c: we need to publish a guide to overall SLSA tooling recommendations with a flowchart of "you have project X in programming language Y in a build environment Z, here's the community recommended approach to achieving SLSA level N". 

While I agree with @MarkLodato that the intro isn't crisp in comparing and contrasting the approaches, I don't think this blog post is really the place to articulate all options in a detailed way. I think this blog should cover why does this (new) approach exist, who is the target audience, and what are its benefits with a brief example showing its use.

Author: FWIW, we have recently discussed exactly what @bobcallaway described with @olivekl related to the documentation to better explain the differences between the different workflows, so that's already very much our intention.

Reviewer 6: Yes, @bobcallaway, agreed -- we discussed just last week the need to update the docs to better direct users between the different options. 

As for the second sentence in the blog post, @MarkLodato, are you looking for something that more explicitly lays out the previous friction points, like:
"This free tool automatically generates SLSA-compliant provenance statements for containers, eliminating  <the room for error? some type of friction?> that container producers experienced with other methods"?

Reviewer: > I think this blog should cover why does this (new) approach exist, who is the target audience, and what are its benefits with a brief example showing its use.

Right. IMO, this blog post does not yet do that. The target audience is someone who already knows the generic generator exists (since it was a prior blog post), so "why does this new approach exist" and "what are its benefits" are relative to that. Without explaining that, it leaves the reader wondering, "Isn't that what the previous post just said?"

> As for the second sentence in the blog post, @MarkLodato, are you looking for something that more explicitly lays out the previous friction points, like:
> "This free tool automatically generates SLSA-compliant provenance statements for containers, eliminating <the room for error? some type of friction?> that container producers experienced with other methods"?

Yup.

Reviewer 7: Some thoughts why this container workflow exists:
1. The format for signing images was pioneered by cosign / sigstore. We can think of it as an ecosystem-specific requirement.
2. A container image lives in a registry, not in an asset of a GitHub release, so different "rights" / ACLs / passwords
3. The provenance for a container also lives on the registry for simplicity (as pioneered by cosign),  so we need the additional inputs (password, etc)
4. Different registries need different code: GCP, AWS, Docker, etc. This code is ecosystem-specific.

The generic generators are intended for blobs / files. We could have squeezed the container feature in it, but it was cleaner to have a dedicated one - due to the difference in inputs needed (username / pwd for registry, etc), attestation format, attestation location, etc. 

It's on us for not naming the first "generic" generator more precisely, like `file-generator`.

Author: > "This free tool automatically generates SLSA-compliant provenance statements for containers, eliminating <the room for error? some type of friction?> that container producers experienced with other methods"?

I tried to do this with the sentence "With the Container Generator these provenance statements can be distributed alongside your images in a container registry and integrate directly with [Sigstore](https://www.sigstore.dev/)-compatible tooling for inspection and verification." as this was specifically what wasn't possible with the previous tools like the generic generator. I updated this a bit to explicitly call out that this was not possible with previous tools as they only supported file artifacts.

Reviewer: Good enough. Thanks.

In only a few short years since container management platforms like Docker and Kubernetes were released, container images have become one of the most popular ways to deploy software. Containers allow their users to deploy software in a more flexible way without depending on language specific packaging or deployment mechanisms. Many public repositories of container images, like [Docker Hub](https://hub.docker.com/), allow you to upload pre-packaged images that can be used out of the box or customized by users as base images.

However, it can be a huge challenge for users to verify the integrity of your images. Various reports from [Aqua Security's Team Nautilus](https://blog.aquasec.com/supply-chain-threats-using-container-images) and the [Sysdig Threat Research Team](https://sysdig.com/blog/analysis-of-supply-chain-attacks-through-public-docker-images/) show that container images containing malicious payloads are common. How can users of your image be certain of how the images you use were built? How can they be sure they haven't been modified to include malicious payloads?

> One solution is including [SLSA provenance](https://slsa.dev/provenance/v0.1) with your container image. SLSA provenance is metadata about your build that includes information that strongly links your artifacts to the build system and source code that was used. SLSA provenance can build trust in your images by helping your users verify that the images come from you and your source repositories.

Author: Done.

However, unlike simply signing your container directly, using the Container Generator as a [GitHub Actions reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) generates the signed provenance in an isolated build job. This protects the provenance from being modified by malicious code in the build itself and enables you to meet the [Non-Falsifiable requirements](https://slsa.dev/spec/v0.1/requirements#non-falsifiable) for [SLSA level 3](https://slsa.dev/spec/v0.1/levels). You can find out more about the specifics in our [Technical Design](https://github.com/slsa-framework/slsa-github-generator/blob/main/SPECIFICATIONS.md).

The following is an example of how to generate provenance using the Container Generator. You can read more details in the [documentation](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container).

### Step 1: Build and push your container image

You can first build and push your image as you would normally using Docker's suite of GitHub Actions.

> ```yaml

Author: I shortened the snippet so hopefully it's not so distracting.

build:
  # ...
  outputs:

>     image: ${{ steps.image.outputs.image }}

Reviewer: these `{{ ... }}` aren't getting rendered, probably because they are interpreted as Jekyll liquid template.

Author: I had this issue before. Fixed.

is_guest_post: false
---

> Previously we announced the [release of the Generic Generator for GitHub Actions](https://slsa.dev/blog/2022/08/slsa-github-workflows-generic-ga). This allowed projects developed in any programming language to generate provenance for artifacts that take the form of regular files. Today, we are announcing the general availability of the [SLSA3 Container Generator for GitHub Actions](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container). This free tool allows any GitHub project to produce SLSA compliant provenance statements so users can verify the origin of container images they use.

Author: Done.
